cdap-sentry/cdap-sentry-extension/cdap-sentry-binding/src/main/java/co/cask/cdap/security/authorization/sentry/binding/AuthBinding.java

```diff
@@ -103,11 +107,111 @@
   private final String instanceName;
   private final String sentryAdminGroup;
 
-  AuthBinding(String sentrySite, String instanceName, String sentryAdminGroup) {
+  private final LoadingCache<Principal, Set<String>> groupCache;
+  private final LoadingCache<String, Set<Role>> roleCache;
+  private final LoadingCache<Role, Map<EntityId, Set<Action>>> entityActionsCache;
+
+  AuthBinding(String sentrySite, final String instanceName, final String sentryAdminGroup,
+              int cacheTtlSecs, int cacheMaxEntries) {
     this.authConf = initAuthzConf(sentrySite);
     this.instanceName = instanceName;
     this.authProvider = createAuthProvider();
     this.sentryAdminGroup = sentryAdminGroup;
+
+    groupCache = CacheBuilder.newBuilder()
+      .expireAfterWrite(cacheTtlSecs, TimeUnit.SECONDS)
+      .maximumSize(cacheMaxEntries)
+      .build(new CacheLoader<Principal, Set<String>>() {
+        @SuppressWarnings("NullableProblems")
+        @Override
+        public Set<String> load(Principal principal) throws Exception {
+          LOG.trace("Group cache miss for principal {}", principal);
+          return getGroups(principal);
+        }
+      });
+
+    roleCache = CacheBuilder.newBuilder()
+      .expireAfterWrite(cacheTtlSecs, TimeUnit.SECONDS)
+      .maximumSize(cacheMaxEntries)
+      .build(new CacheLoader<String, Set<Role>>() {
+        @SuppressWarnings("NullableProblems")
+        @Override
+        public Set<Role> load(final String group) throws Exception {
+          LOG.trace("Role cache miss for group {}", group);
+          return getRoles(group);
+        }
+      });
+
+    entityActionsCache = CacheBuilder.newBuilder()
+      .expireAfterWrite(cacheTtlSecs, TimeUnit.SECONDS)
+      .maximumSize(cacheMaxEntries)
+      .build(new CacheLoader<Role, Map<EntityId, Set<Action>>>() {
+        @SuppressWarnings("NullableProblems")
+        @Override
+        public Map<EntityId, Set<Action>> load(final Role role) throws Exception {
+          LOG.trace("Privilege cache miss for role {}", role);
+          return fetchEntityActions(role);
+        }
+      });
+  }
+
+  private Set<String> getGroups(Principal principal) {
+    return authProvider.getGroupMapping().getGroups(principal.getName());
+  }
+
+  private Set<Role> getRoles(final String group) throws Exception {
```

Comment:
There is another `getRoles` method at Line 571, which used to get the roles for any kind of principal. Is it better to call this method `getRolesForGroup` to avoid confusion?